# clean missing dtypes — exercises

14 tasks that test what the cheat sheet `02_clean_missing_dtypes.ipynb` covers. Try each one
**before** looking at the cheat sheet.

How it works:
1. Read the task. Write your code in the cell below it, assigning the result to the
   variable the task names (usually `answer`).
2. Run the cell: the last line `qN.check()` tells you ✅ or ❌ with a short reason.
3. Stuck? Uncomment `qN.hint()` for a nudge, or `qN.solution()` to see the reference code.

In [1]:
import sys; sys.path.append("..")
from quantlearn import load_questions
q1, q2, q3, q4, q5, q6, q7, q8, q9, q10, q11, q12, q13, q14 = load_questions("02_pandas/02_clean_missing_dtypes")

import numpy as np
import pandas as pd

raw = pd.DataFrame({
    "time": ["2023-01-01 02:00", "2023-01-01 00:00", "2023-01-01 01:00", "2023-01-01 02:00", "2023-01-01 04:00"],
    "cons": [30.0, 10.0, 20.0, 31.0, 50.0],
    "temp": [5.0, -999.0, 6.0, 5.5, -999.0],
    "price": ["70.1", "missing", "23.2", "70.5", "115.8"],
    "region": ["GB", "GB", "GB", "GB", "GB"],
})

In [3]:
raw.dtypes

time       object
cons      float64
temp      float64
price      object
region     object
dtype: object

## Task 1

Convert `raw['time']` to timezone-aware UTC timestamps. Assign the Series.

Assign the result to `answer`.

In [5]:
# your solution here
answer = pd.to_datetime(raw["time"], utc=True)

q1.check()

✅ Correct. expected dtype datetime64[ns, UTC] (pass utc=True)


True

In [7]:
# q1.hint()
q1.solution()

Solution for task 1:

    answer = pd.to_datetime(raw["time"], utc=True)


In [8]:
raw

,time,cons,temp,price,region
0,2023-01-01 02:00,30.0,5.0,70.1,GB
1,2023-01-01 00:00,10.0,-999.0,missing,GB
2,2023-01-01 01:00,20.0,6.0,23.2,GB
3,2023-01-01 02:00,31.0,5.5,70.5,GB
4,2023-01-01 04:00,50.0,-999.0,115.8,GB


## Task 2

Parse the time column (UTC), sort the frame by time and reset the index so it runs 0..n-1. Assign the DataFrame.

Assign the result to `answer`.

In [11]:
# your solution here

answer = raw.copy()
answer["time"] = pd.to_datetime(answer["time"], utc=True)

answer.sort_values(by="time", inplace=True)
answer.reset_index(drop=True, inplace=True)

q2.check()

✅ Correct. 


True

In [10]:
answer

,time,cons,temp,price,region
0,2023-01-01 02:00:00+00:00,30.0,5.0,70.1,GB
1,2023-01-01 00:00:00+00:00,10.0,-999.0,missing,GB
2,2023-01-01 01:00:00+00:00,20.0,6.0,23.2,GB
3,2023-01-01 02:00:00+00:00,31.0,5.5,70.5,GB
4,2023-01-01 04:00:00+00:00,50.0,-999.0,115.8,GB


In [12]:
# q2.hint()
q2.solution()

Solution for task 2:

    df = raw.copy()
    df["time"] = pd.to_datetime(df["time"], utc=True)
    answer = df.sort_values("time").reset_index(drop=True)


## Task 3

`df` is sorted. Two rows share the timestamp 02:00 with slightly different values. Keep only the **last** row for each time. Assign the DataFrame (reset the index).

Assign the result to `answer`.

In [22]:
df = raw.copy()
df["time"] = pd.to_datetime(df["time"], utc=True)
df = df.sort_values("time").reset_index(drop=True)

# your solution here
answer = df.drop_duplicates(subset=["time"], keep="last").reset_index(drop=True)

q3.check()

✅ Correct. 


True

In [23]:
answer

,time,cons,temp,price,region
0,2023-01-01 00:00:00+00:00,10.0,-999.0,missing,GB
1,2023-01-01 01:00:00+00:00,20.0,6.0,23.2,GB
2,2023-01-01 02:00:00+00:00,31.0,5.5,70.5,GB
3,2023-01-01 04:00:00+00:00,50.0,-999.0,115.8,GB


In [21]:
# q3.hint()
q3.solution()


Solution for task 3:

    answer = df.drop_duplicates("time", keep="last").reset_index(drop=True)


## Task 4

Convert `df['price']` (text) to numbers so that bad values become NaN. Assign the numeric Series to `answer` and the number of NaN it contains to `n_bad`.

Assign the result to `answer`, `n_bad`.

In [25]:
df = raw.copy()

# your solution here
answer = pd.to_numeric(df["price"], errors="coerce")
n_bad = answer.isna().sum()

q4.check()

✅ Correct. 


True

In [29]:
answer.isna()
# q4.hint()
# q4.solution()

0    False
1     True
2    False
3    False
4    False
Name: price, dtype: bool

## Task 5

`df['temp']` uses -999 as a sentinel. Assign a Series where every value below -100 is NaN and everything else is unchanged.

Assign the result to `answer`.

In [ ]:
df = raw.copy()

# your solution here
answer = ____

q5.check()

In [ ]:
# q5.hint()
# q5.solution()

## Task 6

Assign a **sorted list** of the columns of `df` that are constant (should be dropped).

Assign the result to `answer`.

In [ ]:
df = raw.copy()

# your solution here
answer = ____

q6.check()

In [ ]:
# q6.hint()
# q6.solution()

## Task 7

`s` is hourly consumption indexed by time, with 03:00 missing. Reindex it to the complete hourly grid from 00:00 to 04:00 and assign the reindexed Series to `answer` and the number of missing hours to `n_missing`.

Assign the result to `answer`, `n_missing`.

In [ ]:
idx = pd.to_datetime(["2023-01-01 00:00", "2023-01-01 01:00", "2023-01-01 02:00", "2023-01-01 04:00"], utc=True)
s = pd.Series([10.0, 20.0, 30.0, 50.0], index=idx)

# your solution here
answer = ____
n_missing = ____

q7.check()

In [ ]:
# q7.hint()
# q7.solution()

## Task 8

Forward-fill `s`, but never fill more than **one** consecutive gap (the second NaN in a run must stay NaN).

Assign the result to `answer`.

In [ ]:
s = pd.Series([10.0, np.nan, np.nan, 40.0, np.nan, 60.0])

# your solution here
answer = ____

q8.check()

In [ ]:
# q8.hint()
# q8.solution()

## Task 9

Fill the gaps in `s` by linear interpolation. Assign the Series.

Assign the result to `answer`.

In [ ]:
s = pd.Series([10.0, np.nan, np.nan, 40.0, np.nan, 60.0])

# your solution here
answer = ____

q9.check()

In [ ]:
# q9.hint()
# q9.solution()

## Task 10

Drop the rows of `df` where `cons` **or** `temp` is NaN (a NaN in `note` alone must not drop the row). Assign the DataFrame.

Assign the result to `answer`.

In [ ]:
df = pd.DataFrame({"cons": [1.0, np.nan, 3.0, 4.0], "temp": [5.0, 6.0, np.nan, 8.0], "note": [np.nan, "a", "b", np.nan]})

# your solution here
answer = ____

q10.check()

In [ ]:
# q10.hint()
# q10.solution()

## Task 11

Fill the NaN in `df['value']` with the **median of the same hour** (group by `hour`). Assign the filled Series.

Assign the result to `answer`.

In [ ]:
df = pd.DataFrame({"hour": [0, 0, 0, 1, 1, 1], "value": [10.0, np.nan, 30.0, 100.0, 120.0, np.nan]})

# your solution here
answer = ____

q11.check()

In [ ]:
# q11.hint()
# q11.solution()

## Task 12

Clip `s` to its 5th and 95th percentiles (values below the 5th become the 5th, above the 95th become the 95th). Assign the Series.

Assign the result to `answer`.

In [ ]:
s = pd.Series([1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 100.0])

# your solution here
answer = ____

q12.check()

In [ ]:
# q12.hint()
# q12.solution()

## Task 13

Clean `region`: strip spaces, make the first letter upper-case and the rest lower-case. Assign the Series.

Assign the result to `answer`.

In [ ]:
region = pd.Series([" london", "WALES ", "North", "scotland"])

# your solution here
answer = ____

q13.check()

In [ ]:
# q13.hint()
# q13.solution()

## Task 14

Real data: clean `../data/hourly_power_raw.csv` enough to count the missing hours: parse `time` as UTC, drop duplicate times, and reindex to the full hourly grid from the first to the last timestamp. Assign the number of missing hours as an integer.

Assign the result to `answer`.

In [ ]:
raw_file = pd.read_csv("../data/hourly_power_raw.csv")

# your solution here
answer = ____

q14.check()

In [ ]:
# q14.hint()
# q14.solution()

---
Done? Re-open the cheat sheet for anything you had to look up, then try the next `_tests` notebook.